src/EnergyPlus/BranchNodeConnections.cc

```diff
@@ -669,7 +669,7 @@ void CheckNodeConnections(EnergyPlusData &state, bool &ErrorsFound)
         FluidStreamInletCount.allocate(MaxFluidStream);
         FluidStreamOutletCount.allocate(MaxFluidStream);
         FluidStreamCounts.allocate(MaxFluidStream);
-        NodeObjects.allocate(state.dataBranchNodeConnections->NumOfNodeConnections);
+        NodeObjects.allocate(state.dataBranchNodeConnections->NumOfNodeConnections + 1);
```

Comment:
Partially yes. The actual key to the solution is the new added at Line 694:
```
        NodeObjects(NumObjects) = state.dataBranchNodeConnections->NumOfNodeConnections + 1;
```

That will let the loop check go till the end of useful range in `NodeObjects`. In most cases, the final value of `NumObjects` will be always smaller than `state.dataBranchNodeConnections->NumOfNodeConnections` since there usually always objects that having two or more nodes, which will then be "folded" into the same `NodeObject(i)' element. So `state.dataBranchNodeConnections->NumOfNodeConnections` is usually quite enough as a valid the upper limit. 

The extra allocation (+1) is to ensure in all cases, the allocated length is still enough for all scenarios, since the new Line 694 will indeed do value assignment on an extra element which needs to be assumed to be always allocated. This is usually not a problem for most normal cases where there are objects that having more than 1 nodes so the original allocation length is usually quite enough. However, for some simple corner cases (e.g. in one of the existing unit tests), out-of-bound segment fault could happen if an extra allocation is not added. The `+1` allocation is a needed safety measure for such corner cases.

Reply:
Well, that definitely doesn't make me feel better about that code as a whole :).  I would like to think that we could identify exactly how big this array needs to be based on _something_.... but it seems your fix is better than what is in develop, so this is fine for right now.  It would be good to identify what is missing that is causing this uncertainty, and language such as: "usually quite enough".  Is it because we are trying to allocate this array too early, and we should defer this step until we have the objects set up?  Or is it because this array should grow dynamically each time a new object is identified rather than trying to guess the size up front?  Anyway, for the scope of this defect fix, it seems this change gets it moving, but it would be nice to go further on a follow up PR.

Reply:
I agree that growing dynamically might be the perfect (and neatest) way for memory allocation. However, it is might not be the best for efficiency, given that there would be repeated 1-increament on array size along the way in the loop. So it looks the current way is somehow a balanced mid-point between perfect memory size vs. repeated vector-size increment operations. 

On the other hand, although named like a complex "node object" array/vector, that particular `NodeObjects` array is actually just of a local primitive int type (Array1D int) and the elements in it will be fully de-allocated even before the Check 10 block finishes. This probably already kept the memory impact to a minimal (size and scope wise). I would be more worried if it is indeed a Node "data struct" type array instead.

Reply:
> there would be repeated 1-increament on array size along the way

`std::vector` is intelligent and reserves more memory behind the scenes, dynamically growing without us knowing it, so that repeated `push_back` calls don't all actually cause a new memory allocation, just an adjustment of the apparent size that we see.

All that to say that this is fine, I'm marking it approved.  Thanks for talking it through.